src/util/MongoBinary.js

```diff
@@ -43,84 +122,60 @@ export default class MongoBinary {
       arch: process.env?.MONGOMS_ARCH || os.arch(),
       version: process.env?.MONGOMS_VERSION || 'latest',
+      systemBinary: process.env?.MONGOMS_SYSTEM_BINARY,
       debug:
         typeof process.env.MONGOMS_DEBUG === 'string'
           ? ['1', 'on', 'yes', 'true'].indexOf(process.env.MONGOMS_DEBUG.toLowerCase()) !== -1
           : false,
     };
 
-    let debug;
     if (opts.debug) {
       if (opts.debug.call && typeof opts.debug === 'function' && opts.debug.apply) {
-        debug = opts.debug;
+        this.debug = opts.debug;
       } else {
-        debug = console.log.bind(null);
+        this.debug = console.log.bind(null);
       }
     } else {
-      debug = (msg: string) => {}; // eslint-disable-line
+      this.debug = (msg: string) => {}; // eslint-disable-line
     }
 
     const options = { ...defaultOptions, ...opts };
-    debug(`MongoBinary options: ${JSON.stringify(options)}`);
+    this.debug(`MongoBinary options: ${JSON.stringify(options)}`);
 
-    const { downloadDir, platform, arch, version } = options;
+    const { version, systemBinary } = options;
 
-    if (this.cache[version]) {
-      debug(`MongoBinary: found cached binary path for ${version}`);
-    } else {
-      // create downloadDir if not exists
-      await new Promise((resolve, reject) => {
-        mkdirp(downloadDir, err => {
-          if (err) reject(err);
-          else resolve();
-        });
-      });
+    let binaryPath: string = '';
 
-      const lockfile = path.resolve(downloadDir, `${version}.lock`);
-
-      // wait lock
-      await new Promise((resolve, reject) => {
-        LockFile.lock(
-          lockfile,
-          {
-            wait: 120000,
-            pollPeriod: 100,
-            stale: 110000,
-            retries: 3,
-            retryWait: 100,
-          },
-          err => {
-            if (err) reject(err);
-            else resolve();
-          }
-        );
-      });
+    if (systemBinary) {
+      binaryPath = await this.getSystemPath(systemBinary);
+    }
 
-      // again check cache, maybe other instance resolve it
-      if (!this.cache[version]) {
-        const downloader = new MongoBinaryDownload({
-          downloadDir,
-          platform,
-          arch,
-          version,
-        });
-
-        downloader.debug = debug;
-        this.cache[version] = await downloader.getMongodPath();
-      }
+    if (!binaryPath) {
+      binaryPath = await this.getCachePath(version);
+    }
 
-      // remove lock
-      LockFile.unlock(lockfile, err => {
-        debug(
-          err
-            ? `MongoBinary: Error when removing download lock ${err}`
-            : `MongoBinary: Download lock removed`
-        );
-      });
+    if (!binaryPath) {
+      binaryPath = await this.getDownloadPath(options);
     }
 
-    debug(`MongoBinary: Mongod binary path: ${this.cache[version]}`);
-    return this.cache[version];
+    if (version !== 'latest' && systemBinary) {
```

Comment:
We should not show the message if versions are the same ;)

And I think that better to move this code inside `if (systemBinary) {` at line 149
right after `binaryPath = await this.getSystemPath(systemBinary);`

Reply:
All rest parts are good! Code became much cleaner! Thanks!

Reply:
Ouch! And don't forget to add new option to README.md

Reply:
Thanks, I appreciate the feedback and agree.  I have updated the code and the README.md.  let me know if there is anything else.